code/modules/modular_computers/networking/machinery/acl.dm

```diff
@@ -0,0 +1,208 @@
+
+	if(href_list["change_file_server"])
+		var/list/file_servers = network.get_file_server_tags()
+		var/file_server = input(usr, "Choose a fileserver to view access records on:", "Select File Server") as null|anything in file_servers
+		if(file_server)
+			file_source.server = file_server
+			return TOPIC_REFRESH
+
+	if(href_list["assign_grant"])
+		// Resolve our selection back to a file.
+		var/datum/computer_file/data/grant_record/grant = computer.get_grant(href_list["assign_grant"])
+		if(!grant)
+			error = "ERROR: Grant record not found."
+			return TOPIC_REFRESH
+		var/datum/computer_file/report/crew_record/AR = get_access_record()
+		if(!AR)
+			error = "ERROR: Access record not found."
+			return TOPIC_REFRESH
+		AR.add_grant(grant)
+
+	if(href_list["remove_grant"])
+		var/datum/computer_file/report/crew_record/AR = get_access_record()
+		if(!AR)
+			error = "ERROR: Access record not found."
+			return TOPIC_REFRESH
+		AR.remove_grant(href_list["remove_grant"]) // Add the grant to the record.
+
+	if(href_list["create_grant"])
+		var/new_grant_name = uppertext(sanitize(input(usr, "Enter the name of the new grant:", "Create Grant")))
+		if(!CanInteract(user, GLOB.default_state))
+			return TOPIC_REFRESH
+		if(!new_grant_name)
+			return TOPIC_REFRESH
+		if(!create_grant(new_grant_name))
+			error = "MAINFRAME ERROR: Unable to store grant on mainframe."
+			return TOPIC_REFRESH
+
+	if(href_list["delete_grant"])
+		if(!file_source.delete_file(href_list["delete_grant"]))
+			error = "FIREWALL ERROR: Unable to delete grant."
+			return TOPIC_REFRESH
+
+	if(href_list["view_user"])
+		editing_user = href_list["view_user"]
+
+	if(href_list["write_id"])
+		var/obj/item/stock_parts/computer/card_slot/card_slot = get_component_of_type(/obj/item/stock_parts/computer/card_slot)
+		var/obj/item/card/id/network/card = card_slot.stored_card
+		if(!card)
+			error = "HARDWARE ERROR: No valid card inserted."
+			return TOPIC_REFRESH
+		// Write to the card.
+		var/datum/computer_file/report/crew_record/AR = get_access_record()
+		card.network_id = computer.network_id
+		card.user_id = AR.user_id
+		card.access_record = AR
+		card.broken = FALSE
+		visible_message(SPAN_NOTICE("\The [src] clicks and hums, writing new data to \a [card]."))
+
+	if(href_list["eject_id"])
+		var/obj/item/stock_parts/computer/card_slot/card_slot = get_component_of_type(/obj/item/stock_parts/computer/card_slot)
+		if(!card_slot.stored_card)
+			error = "HARDWARE ERROR: No valid card inserted."
+			return TOPIC_REFRESH
+		card_slot.eject_id(user)
+
+	if(href_list["add_admin"])
+		network.access_controller.add_admin(editing_user)
+
+	if(href_list["remove_admin"])
+		network.access_controller.rem_admin(editing_user)
+
+/obj/machinery/network/acl/ui_data(mob/user, ui_key)
+	. = ..()
+
+	if(error)
+		return
+
+	var/obj/item/stock_parts/computer/card_slot/card_slot = get_component_of_type(/obj/item/stock_parts/computer/card_slot)
+	.["card_inserted"] = !!card_slot.stored_card
```

Comment:
gotta check if slot is present at all

Reply:
honest question: is that necessary if it's a required component for the machine to be made? I was under the assumption that was enough.

Reply:
...but you didn't set it as required, I thought that was intentional?
Either put it in req_components on the circuitboard or handle its lack